### PR TITLE
[ONNX] Add col2im for opset 18

### LINKED ISSUE #84408: Exporting the operator ::col2im to ONNX opset version 11 is not supported
### 🐛 Describe the bug

When I converted the model in “.pt” format to onnx format, I received an error that the operator col2im is not supported.

## code

    import torch
    from cvnets import get_model
    from options.opts import get_segmentation_eval_arguments
    
    def pt2onnx():
        opts = get_segmentation_eval_arguments()
        model = get_model(opts)
        model.eval()
        onnx_save_path = "model/mobilevit.onnx"
        in_data = torch.randn(1, 3, 512, 512)
        torch.onnx.export(model, 
                          in_data, 
                          onnx_save_path, 
                          opset_version=11, 
                          do_constant_folding=True, 
                          input_names=["in"],
                          output_names=["out"])
        return

    if __name__ == '__main__':
        pt2onnx()

## error
Traceback (most recent call last):
  File "/home/sunseeker/project/robot_seg/code/mobilevit_seg/demo.py", line 20, in <module>
    pt2onnx()
  File "/home/sunseeker/project/robot_seg/code/mobilevit_seg/demo.py", line 13, in pt2onnx
    torch.onnx.export(model, in_data, onnx_save_path, opset_version=11, do_constant_folding=True, input_names=["in"],
  File "/opt/anaconda3/envs/mobilevit/lib/python3.10/site-packages/torch/onnx/__init__.py", line 350, in export
    return utils.export(
  File "/opt/anaconda3/envs/mobilevit/lib/python3.10/site-packages/torch/onnx/utils.py", line 163, in export
    _export(
  File "/opt/anaconda3/envs/mobilevit/lib/python3.10/site-packages/torch/onnx/utils.py", line 1074, in _export
    graph, params_dict, torch_out = _model_to_graph(
  File "/opt/anaconda3/envs/mobilevit/lib/python3.10/site-packages/torch/onnx/utils.py", line 731, in _model_to_graph
    graph = _optimize_graph(
  File "/opt/anaconda3/envs/mobilevit/lib/python3.10/site-packages/torch/onnx/utils.py", line 308, in _optimize_graph
    graph = _C._jit_pass_onnx(graph, operator_export_type)
  File "/opt/anaconda3/envs/mobilevit/lib/python3.10/site-packages/torch/onnx/__init__.py", line 416, in _run_symbolic_function
    return utils._run_symbolic_function(*args, **kwargs)
  File "/opt/anaconda3/envs/mobilevit/lib/python3.10/site-packages/torch/onnx/utils.py", line 1421, in _run_symbolic_function
    raise symbolic_registry.UnsupportedOperatorError(
**torch.onnx.symbolic_registry.UnsupportedOperatorError: Exporting the operator ::col2im to ONNX opset version 11 is not supported. Please feel free to request support or submit a pull request on PyTorch GitHub.**

## ENV
PyTorch version: 1.12.1
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (Ubuntu 11.2.0-19ubuntu1) 11.2.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.4 (main, Mar 31 2022, 08:41:55) [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.15.0-47-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3050
Nvidia driver version: 510.85.02
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.23.1
[pip3] pytorchvideo==0.1.5
[pip3] torch==1.12.1
[pip3] torchaudio==0.12.1
[pip3] torchvision==0.13.1
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               10.2.89              hfd86e86_1  
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2021.4.0           h06a4308_640  
[conda] mkl-service               2.4.0           py310h7f8727e_0  
[conda] mkl_fft                   1.3.1           py310hd6ae3a3_0  
[conda] mkl_random                1.2.2           py310h00e6091_0  
[conda] numpy                     1.23.1          py310h1794996_0  
[conda] numpy-base                1.23.1          py310hcba007f_0  
[conda] pytorch                   1.12.1          py3.10_cuda10.2_cudnn7.6.5_0    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] pytorchvideo              0.1.5                    pypi_0    pypi
[conda] torchaudio                0.12.1              py310_cu102    pytorch
[conda] torchvision               0.13.1              py310_cu102    pytorch

### PR DESCRIPTION
**Pull request by thiagocrepaldi**
Opset 18 will be used to introduce suport for ONNX's Col2Im-18 and resolve https://github.com/pytorch/pytorch/issues/84408

Depends: https://github.com/pytorch/pytorch/pull/83201 (CI will fail until ONNX submodule is updated)

as per Faith recommendation, this PR should be merged post ORT 1.13 only

### CODE REVIEW

**Comment by justinchuby** [comment]
Looks like the ort version we use doesn't support opset 18 yet. Can we skip tests for opset 18 for now?

**Review by justinchuby — APPROVED** [review]

File: torch/onnx/_constants.py
Comment (justinchuby) [inline torch/onnx/_constants.py:5]: We can consider adding an "experimental_opset"

Comment (thiagocrepaldi) [reply torch/onnx/_constants.py]: The root cause for the failure is that ORT does not support opset 18. a new `onnx_next_opset = 18` wouldn't help PyTorch itself or any other engine, but ORT only. So a new constant here would be dead weight.

Better keep using 17 until I push a PR to ONNX Runtime setting 18 as the max opset. That could happen if we change https://github.com/microsoft/onnxruntime/pull/12311 from contrib to ONNX.

Comment (justinchuby) [reply torch/onnx/_constants.py]: Set as experimental, we can enable behaviors like warning users that it is a version that is not released?

**Comment by BowenBao** [comment]
IIUC, depends on a bump on `third_party/onnx` version, and even #83201 is not current enough.

File: torch/onnx/symbolic_opset18.py
Comment (thiagocrepaldi) [reply torch/onnx/symbolic_opset18.py:27]: good catch

File: test/onnx/test_pytorch_onnx_no_runtime.py
Comment (AllenTiTaiWang) [inline test/onnx/test_pytorch_onnx_no_runtime.py:783]: I think it'd better we set a range [0, 255] if you are simulating image here?

File: test/onnx/test_pytorch_onnx_no_runtime.py
Comment (AllenTiTaiWang) [inline test/onnx/test_pytorch_onnx_no_runtime.py:789]: Curious about if we will need a `im2col` Op after this? Or we plan to use not col2im?

**Review by titaiwangms — APPROVED** [review]
LGTM!

File: test/onnx/test_pytorch_onnx_no_runtime.py
Comment (justinchuby) [inline test/onnx/test_pytorch_onnx_no_runtime.py:805]: Prefer assertEqual and other more specific assertions so we get rich errors when they fail.

File: torch/onnx/symbolic_opset18.py
Comment (justinchuby) [inline torch/onnx/symbolic_opset18.py:25]: Move bear type to closet to the function (bottom) so it checks the annotations properly

File: torch/onnx/symbolic_opset18.py
Comment (justinchuby) [inline torch/onnx/symbolic_opset18.py:27]: Annotate types?

File: torch/onnx/symbolic_opset18.py
Comment (justinchuby) [inline torch/onnx/symbolic_opset18.py:28]: List comprehension with two for loops can be consider too complicated to reason about. Consider writing out as for loops?

https://google.github.io/styleguide/pyguide.html#274-decision

File: torch/onnx/symbolic_opset18.py
Comment (justinchuby) [inline torch/onnx/symbolic_opset18.py:33]: For things like dilations, if we are passing them right through, I wonder if we should avoid parsing them and recreating constant nodes for efficiency.

File: test/onnx/test_pytorch_onnx_no_runtime.py
Comment (thiagocrepaldi) [reply test/onnx/test_pytorch_onnx_no_runtime.py:783]: Can you elaborate the rationale?
The image size is not that important, we just need one image with known result.
I didnt use same values for for dilation, padding and stride (instead of `1` for everything) to make sure to result would not be a false positive

File: test/onnx/test_pytorch_onnx_no_runtime.py
Comment (thiagocrepaldi) [reply test/onnx/test_pytorch_onnx_no_runtime.py:789]: [We have already,](https://github.com/pytorch/pytorch/blob/master/torch/onnx/symbolic_opset11.py#L1111) we have it already. im2col cannot be used to express col2im and vice-versa, though

File: test/onnx/test_pytorch_onnx_no_runtime.py
Comment (thiagocrepaldi) [reply test/onnx/test_pytorch_onnx_no_runtime.py:805]: done

File: test/onnx/test_pytorch_onnx_no_runtime.py
Comment (AllenTiTaiWang) [reply test/onnx/test_pytorch_onnx_no_runtime.py:783]: I was referring that pixel range is [0, 255], not the image size.

File: torch/onnx/symbolic_opset18.py
Comment (thiagocrepaldi) [reply torch/onnx/symbolic_opset18.py:25]: Thanks, do we have this instruction in the code somewhere?

Comment (justinchuby) [reply torch/onnx/symbolic_opset18.py:25]: I can add that in readme.

File: torch/onnx/symbolic_opset18.py
Comment (thiagocrepaldi) [reply torch/onnx/symbolic_opset18.py:33]: If we detect considerable slowdowns because of it, we could do it.
However, IMO simplicity and readability might be better qualities for non-critical pieces of code. The lesser unnecessary code we write, the fewer bugs we may introduce :)

Comment (justinchuby) [reply torch/onnx/symbolic_opset18.py:33]: SGTM

File: test/onnx/test_pytorch_onnx_no_runtime.py
Comment (thiagocrepaldi) [reply test/onnx/test_pytorch_onnx_no_runtime.py:783]: I still don't get it. What would be the code after your proposal? That might help me get your point

File: torch/onnx/symbolic_opset18.py
Comment (thiagocrepaldi) [reply torch/onnx/symbolic_opset18.py:25]: Is there a way the order is not important? Maybe creating a base class for all decorators that work as a stack context manager and the bear type is automatically added in the right spot?

I missed this order quite easily, so I wonder if that will not happen again with another person

Comment (justinchuby) [reply torch/onnx/symbolic_opset18.py:25]: That works. We may not even need the context manager thing. decorators are just higher order functions

Comment (justinchuby) [reply torch/onnx/symbolic_opset18.py:25]: That said, it also helps to think about the order when decorators are applied. Having it explicit has its advantage, making what we apply intentional

File: torch/onnx/symbolic_opset18.py
Comment (thiagocrepaldi) [reply torch/onnx/symbolic_opset18.py]: Good catch, I got hypothesized by the `parse_args` and forgot about it lol

File: torch/onnx/symbolic_opset18.py
Comment (thiagocrepaldi) [reply torch/onnx/symbolic_opset18.py]: makes sense

File: torch/onnx/symbolic_opset18.py
Comment (justinchuby) [inline torch/onnx/symbolic_opset18.py:37]: Sequence

File: torch/onnx/symbolic_opset18.py
Comment (justinchuby) [inline torch/onnx/symbolic_opset18.py:45]: Does this mean it does it twice?

```suggestion
adjusted_padding.extend(padding).extend(padding)
```

or
```
adjusted_padding = list(padding) * 2
```
?

Comment (thiagocrepaldi) [reply torch/onnx/symbolic_opset18.py:45]: not quite, but there was a bug on my code.

**Review by justinchuby — APPROVED** [review]

**Comment by justinchuby** [comment]
Please wait for #84448 to merge. After which you will be able to use the decorator for registering the symbolic function

File: torch/onnx/symbolic_opset18.py
Comment (justinchuby) [inline torch/onnx/symbolic_opset18.py:28]: We need to add col2im to `__all__`

File: torch/onnx/symbolic_opset18.py
Comment (justinchuby) [inline torch/onnx/symbolic_opset18.py:1]: We need to import this file in `__init__` so the registry picks it up

File: torch/onnx/symbolic_opset18.py
Comment (justinchuby) [inline torch/onnx/symbolic_opset18.py:32]: I'd add the decorator to register this function

**Review by justinchuby — CHANGES_REQUESTED** [review]
minor comments

File: torch/onnx/symbolic_opset18.py
Comment (thiagocrepaldi) [reply torch/onnx/symbolic_opset18.py:28]: Isn't `__all__` automatically populated? We only customize it when we want to remove something from it. 
Which change caused this extra step to be needed? It is hard to keep static lists up to date

Comment (justinchuby) [reply torch/onnx/symbolic_opset18.py:28]: unfortunately this is how pytorch maintainer suggested https://github.com/pytorch/pytorch/wiki/Public-API-definition-and-documentation. We needed all because we are also importing from typing (which we want to do) and the classes from typing "appears" to be public

Comment (justinchuby) [reply torch/onnx/symbolic_opset18.py:28]: After adding this file to init I would expect a test error that says this

Comment (justinchuby) [reply torch/onnx/symbolic_opset18.py:28]: this will be temporary though. After we take the symbolic functions into internal we will have a programmatic way of populating all

File: torch/onnx/symbolic_opset18.py
Comment (justinchuby) [inline torch/onnx/symbolic_opset18.py:31]: The partial `_onnx_symbolic` like in other opset files

**Review by justinchuby — APPROVED** [review]

File: test/onnx/test_pytorch_onnx_no_runtime.py
Comment (abock) [inline test/onnx/test_pytorch_onnx_no_runtime.py:779]: Can you create a tracking issue for this? Otherwise it risks being forgotten about.

**Review by abock — APPROVED** [review]

File: test/onnx/test_pytorch_onnx_no_runtime.py
Comment (thiagocrepaldi) [reply test/onnx/test_pytorch_onnx_no_runtime.py:779]: Tracking through https://github.com/pytorch/pytorch/issues/84408

**Review by BowenBao — APPROVED** [review]

**Comment by thiagocrepaldi** [comment]
> @thiagocrepaldi : Is this going to be merged soon? This op is needed to export NVIDIA's RadTTS network.

Hello @borisfom We are waiting on Meta to fix proprietary code that rely on older ONNX to merge this. 
I am hoping this will be merged sometime in January/2023

**Comment by borisfom** [comment]
Thanks! We may be able to work around our particular use case with reshape/transpose - but there is great interest in general to be on top of ONNX spec.

**Pull request closed without merging**
